File: polyglot-utils/php/preact-iso-url-pattern.php
<?php
// Run program: php preact-iso-url-pattern.php

// Safe URL decode function with error handling
function safeUrldecode($str) {
    if ($str === null || $str === '') {
        return $str;
    }
    
    // urldecode in PHP generally doesn't throw exceptions,
    // but we can add validation for malformed percent encoding
    $decoded = urldecode($str);
    
    // If the original contained a % but decoding didn't change much,
    // it might be malformed, but PHP's urldecode is quite tolerant
    return $decoded;
}

function preactIsoUrlPatternMatch($url, $route, $matches = null) {
    if ($matches === null) {
        $matches = ['params' => (object)[]];
    }
    $url = array_values(array_filter(explode('/', $url)));
    $route = array_values(array_filter(explode('/', $route ?? '')));

    for ($i = 0; $i < max(count($url), count($route)); $i++) {
        preg_match('/^(:?)(.*?)([+*?]?)$/', $route[$i] ?? '', $parts);
        $m = $parts[1] ?? '';
        $param = $parts[2] ?? '';
        $flag = $parts[3] ?? '';
        $val = $url[$i] ?? null;

        // segment match:
        if (!$m && $param === $val) continue;
        
        // /foo/* match
        if (!$m && $val && $flag == '*') {
            $decodedParts = array_map('safeUrldecode', array_slice($url, $i));
            $matches['rest'] = '/' . implode('/', $decodedParts);
            break;
        }

        // segment mismatch / missing required field:
        if (!$m || (!$val && $flag != '?' && $flag != '*')) {
            return null;
        }
        $rest = $flag == '+' || $flag == '*';

        // rest (+/*) match:
        if ($rest) {
            $decodedParts = array_map('safeUrldecode', array_slice($url, $i));
            $val = implode('/', $decodedParts) ?: null;
        }
        // normal/optional field:
        elseif ($val) {
            $val = safeUrldecode($url[$i]);
        }

        $matches['params']->$param = $val;
        if (!isset($matches[$param])) {
            $matches[$param] = $val;
        }

        if ($rest) break;
    }

    return $matches;
}
// Example usage:
// var_dump(preactIsoUrlPatternMatch("/foo/bar%20baz", "/foo/:param"));
// var_dump(preactIsoUrlPatternMatch("/foo/bar/baz", "/foo/*"));
// var_dump(preactIsoUrlPatternMatch("/foo", "/foo/:param?"));
// var_dump(preactIsoUrlPatternMatch("/foo/bar", "/bar/:param"));
// var_dump(preactIsoUrlPatternMatch('/users/test%40example.com/posts', '/users/:userId/posts'));
?>

File: polyglot-utils/go/preact_iso_url_pattern.go
// Run program: go run preact-iso-url-pattern.go

package main

import (
	// "fmt"
	"net/url"
	"regexp"
	"strings"
)

type Matches struct {
	Params map[string]string `json:"params"`
	Rest   string            `json:"rest,omitempty"`
}

func preactIsoUrlPatternMatch(urlStr, route string, matches *Matches) *Matches {
	if matches == nil {
		matches = &Matches{
			Params: make(map[string]string),
		}
	}
	urlParts := filterEmpty(strings.Split(urlStr, "/"))
	routeParts := filterEmpty(strings.Split(route, "/"))

	for i := 0; i < max(len(urlParts), len(routeParts)); i++ {
		var m, param, flag string
		if i < len(routeParts) {
			re := regexp.MustCompile(`^(:?)(.*?)([+*?]?)$`)
			matches := re.FindStringSubmatch(routeParts[i])
			if len(matches) > 3 {
				m, param, flag = matches[1], matches[2], matches[3]
			}
		}

		var val string
		if i < len(urlParts) {
			val = urlParts[i]
		}

		// segment match:
		if m == "" && param == val {
			continue
		}

		// /foo/* match
		if m == "" && val != "" && flag == "*" {
			matches.Rest = "/" + strings.Join(urlParts[i:], "/")
			break
		}

		// segment mismatch / missing required field:
		if m == "" || (val == "" && flag != "?" && flag != "*") {
			return nil
		}

		rest := flag == "+" || flag == "*"

		// rest (+/*) match:
		if rest {
			decodedParts := make([]string, len(urlParts[i:]))
			for j, part := range urlParts[i:] {
				decoded, err := url.QueryUnescape(part)
				if err != nil {
					decoded = part // fallback to original if decode fails
				}
				decodedParts[j] = decoded
			}
			val = strings.Join(decodedParts, "/")
		} else if val != "" {
			// normal/optional field: decode val (like JavaScript does)
			decoded, err := url.QueryUnescape(val)
			if err != nil {
				decoded = urlParts[i]
			}
			val = decoded
		}

		matches.Params[param] = val

		if rest {
			break
		}
	}

	return matches
}

func filterEmpty(s []string) []string {
	var result []string
	for _, str := range s {
		if str != "" {
			result = append(result, str)
		}
	}
	return result
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

// Example usage:
// func main() {
// 	params := &Matches{Params: make(map[string]string)}
// 	fmt.Println(preactIsoUrlPatternMatch("/foo/bar%20baz", "/foo/:param", params))
//
// 	params := &Matches{Params: make(map[string]string)}
// 	fmt.Println(preactIsoUrlPatternMatch("/foo/bar/baz", "/foo/*"))
//
// 	params := &Matches{Params: make(map[string]string)}
// 	fmt.Println(preactIsoUrlPatternMatch("/foo", "/foo/:param?"))
//
// 	params := &Matches{Params: make(map[string]string)}
// 	fmt.Println(preactIsoUrlPatternMatch("/foo/bar", "/bar/:param"))
//
// 	params := &Matches{Params: make(map[string]string)}
// 	fmt.Println(preactIsoUrlPatternMatch("/users/test%40example.com/posts", "/users/:userId/posts"))
// }
